Enhancement: add image uploader settings import/export functionality

knobby-tires opened the pull request:
Adds import/export functionality for image uploader settings as requested in #6283.

pajlada commented:
Hi!
At a quick glance through the code, things look good. I'll give it a proper review & test this weekend.
Thanks for your contribution!

knobby-tires commented:
Awesome! I am leaving town this weekend and will have very limited access to a computer. I am happy to make necessary adjustments when/if I can. Just want to make sure you know I am not abandoning the PR lol.

pajlada reviewed (changes requested):
As a first pass - will do another more in-depth review once it's more testable

When I was reviewing this I was noticing the text fields don't update when the setting updates from external sources - I'll be fixing that in a separate PR.

knobby-tires commented:
Hi @pajlada, I believe I addressed most of what you asked :

- I moved import/export functions to `ImageUploader.hpp/cpp` in the `chatterino::imageuploader::detail` namespace
- I changed `exportSettings` to take `Settings &s` parameter as suggested: `QJsonObject exportSettings(Settings &s);`
- I moved all related functions (`importSettings`, `validateImportJson`, `parseAndApplyHeaders`) to the new file
- Updated `ExternalToolsPage` to use the new functions
- Added the new files to `CMakeLists.txt`
- Confirmed functionality still works correctly (export/import both working)


Next I'll work on writing unit tests for the `exportSettings` function next 

Thanks for the detailed feedback, if I have misunderstood anything up until now please let me know. I am really trying to learn and gain real experience here.

pajlada commented on src/util/ImageUploader.cpp:
```diff
@@ -0,0 +1,118 @@
+    return settingsObj;
+}
+
+bool importSettings(const QJsonObject &settingsObj, Settings &s)
```
Question:
This function would currently return true if the user _only_ has "URL" set.
Is the constraint that only one of the fields need to be set, or that all fields need to be set?

pajlada commented on src/util/ImageUploader.hpp:
```diff
@@ -0,0 +1,20 @@
+namespace imageuploader {
+namespace detail {
+
+QJsonObject exportSettings(Settings &s);
```
Make this take a `const Settings &s` instead, to clarify that we're not looking to make any modifications to the settings in this function.

pajlada commented on src/util/ImageUploader.hpp:
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+class QJsonObject;
```
You're missing a forward declaration to `QString` here

pajlada commented on src/util/ImageUploader.hpp:
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
It would be nice with a tiny comment for each of the 4 functions you've made. It does not have to be a long comment, but just a tiny insight into what it does (e.g. "Exports settings from Chatterino to the ShareX export JSON format" or something)

pajlada commented on src/util/ImageUploader.hpp:
```diff
@@ -0,0 +1,20 @@
+QJsonObject exportSettings(Settings &s);
+bool importSettings(const QJsonObject &settingsObj, Settings &s);
+bool validateImportJson(const QString &clipboardText, QJsonObject &settingsObj);
+void parseAndApplyHeaders(const QJsonObject &headersObj, Settings &s);
```
This function is only ever used by the internal pieces - you can make it a source-only function.
Additionally, it would be nice if it only did the parsing, letting the `importSettings` function do the importing.
I imagine the signature would change to something like
```c++
QStringList parseHeaders(const QJsonObject &headersObj);
```

pajlada commented on src/widgets/settingspages/ExternalToolsPage.hpp:
```diff
@@ -14,8 +14,10 @@ class ExternalToolsPage : public SettingsPage
 
 private:
     void initLayout(GeneralPageView &layout);
+    void exportSettings();
+    void importSettings();
```
I'd prefer if these functions were source-only, and accepted a `QWidget *parent` parameter instead of using `this` for the dialog parents. You can still pass `this` to them. They should also be renamed to clarify that they are specifically importing or exporting image uploader settings, not any generic settings.

pajlada commented on src/widgets/settingspages/ExternalToolsPage.cpp:
```diff
@@ -135,9 +143,73 @@ void ExternalToolsPage::initLayout(GeneralPageView &layout)
+
+void ExternalToolsPage::importSettings()
+{
+    QString clipboardText = QApplication::clipboard()->text().trimmed();
```
Use our `getClipboardText` function (so `getClipboardText().trimmed()`)

pajlada commented on src/widgets/settingspages/ExternalToolsPage.cpp:
```diff
@@ -135,9 +143,73 @@ void ExternalToolsPage::initLayout(GeneralPageView &layout)
+
+    QJsonObject settingsObj = imageuploader::detail::exportSettings(s);
+    QJsonDocument doc(settingsObj);
+    QApplication::clipboard()->setText(doc.toJson(QJsonDocument::Indented));
```
Use our `crossPlatformCopy` function (so `crossPlatformCopy(doc.toJson(QJsonDocument::Indented));`

pajlada commented on src/util/ImageUploader.cpp:
```diff
@@ -0,0 +1,118 @@
+
+bool importSettings(const QJsonObject &settingsObj, Settings &s)
+{
+    bool hasValidSettings = false;
```
Are there any "meta" fields that _always_ need to be part of the export that we can check?
e.g. if the json blob always needs `Name` and `Version`, we could check those here and error out early if something is wrong

pajlada replied:
Only the meta fields from ShareX exports, which I think are `Version` and `Name`

pajlada reviewed (changes requested):
File structure looks good from my perspective.
Some nits with how functions are structured, some documentation, and some questions regarding validation.

knobby-tires commented:
Hey, I addressed your feedback from yesterday. Let me know what you think

pajlada reviewed (approved):
This looks good to me now - I took the liberty of doing a few minor modifications.
Could you take a look at the changes I made and see if they make sense to you @knobby-tires ?

knobby-tires commented:
@pajlada, Looks good to me! Thanks for the help.
